packages/bff/src/services/toolsService.ts

```diff
@@ -0,0 +1,395 @@
+  verifyClientAssertion,
+  verifyClientAssertionSignature,
+} from "pagopa-interop-client-assertion-validation";
+import {
+  AgreementId,
+  ApiError,
+  ClientId,
+  EServiceId,
+  ItemState,
+  PurposeId,
+  TenantId,
+  unsafeBrandId,
+} from "pagopa-interop-models";
+import { getAllFromPaginated, WithLogger } from "pagopa-interop-commons";
+import {
+  agreementApi,
+  authorizationApi,
+  bffApi,
+  catalogApi,
+  purposeApi,
+} from "pagopa-interop-api-clients";
+import {
+  AgreementProcessClient,
+  CatalogProcessClient,
+  PagoPAInteropBeClients,
+} from "../providers/clientProvider.js";
+import { BffAppContext } from "../utilities/context.js";
+import {
+  activeAgreementByEserviceAndConsumerNotFound,
+  clientAssertionPublicKeyNotFound,
+  ErrorCodes,
+  eserviceDescriptorNotFound,
+  missingActivePurposeVersion,
+  multipleAgreementForEserviceAndConsumer,
+  purposeIdNotFoundInClientAssertion,
+} from "../model/domain/errors.js";
+import { TokenGenerationValidationStepFailure } from "../../../api-clients/dist/bffApi.js";
+
+export function toolsServiceBuilder(clients: PagoPAInteropBeClients) {
+  return {
+    async validateTokenGeneration(
+      clientId: string | undefined,
+      clientAssertion: string,
+      _clientAssertionType: string,
+      _grantType: string,
+      ctx: WithLogger<BffAppContext>
+    ): Promise<bffApi.TokenGenerationValidationResult> {
+      ctx.logger.info(`Validating token generation for client ${clientId}`);
+
+      const { errors: parametersErrors } = validateRequestParameters({
+        client_assertion: clientAssertion,
+        client_assertion_type:
+          "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+        grant_type: "client_credentials",
+        client_id: clientId,
+      });
+
+      const { data: jwt, errors: clientAssertionErrors } =
+        verifyClientAssertion(clientAssertion, clientId);
+
+      if (parametersErrors || clientAssertionErrors) {
+        return handleValidationResults({
+          clientAssertionErrors: [
+            ...(parametersErrors ?? []),
+            ...(clientAssertionErrors ?? []),
+          ],
+        });
+      }
+
+      const { data: key, errors: keyRetrieveErrors } = await retrieveKey(
+        clients,
+        jwt,
+        ctx
+      );
+
+      if (keyRetrieveErrors) {
+        return handleValidationResults({
+          keyRetrieveErrors,
+        });
+      }
+
+      const { errors: clientSignatureErrors } = verifyClientAssertionSignature(
+        clientAssertion,
+        key
+      );
+
+      if (clientSignatureErrors) {
+        return handleValidationResults({
+          clientSignatureErrors,
+        });
+      }
+
+      const { errors: platformStateErrors } =
+        validateClientKindAndPlatformState(key, jwt);
+
+      if (platformStateErrors) {
+        return handleValidationResults({
+          platformStateErrors,
+        });
+      }
+
+      return handleValidationResults({});
```

Comment:
🤔  why empty body ? In scala code use build a [TokenGenerationValidationResult](https://vscode.dev/github/pagopa/interop-be-backend-for-frontend/blob/1.0.x/src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/ToolsApiServiceImpl.scala#L134) providing `clientKind` and `eService`.

Reply:
Yes it was still a WIP! i now implemented the eservice retrieval and returned clientKind and eservice where needed as in scala code.